Mathlib/Analysis/Polynomial/MahlerMeasure.lean

```diff
@@ -137,8 +140,114 @@ theorem prod_mahlerMeasure_eq_mahlerMeasure_prod (s : Multiset ℂ[X]) :
+  · simp
+  have hmeroOnB : MeromorphicOn (· - z) (closedBall 0 |1|) := id.sub <| const z
+  simp_rw [logMahlerMeasure_def, eval_sub, eval_X, eval_C]
+  rw [circleAverage_log_norm one_ne_zero hmeroOnB, abs_one,
+  -- get rid of the `meromorphicTrailingCoeffAt`
+    AnalyticAt.meromorphicTrailingCoeffAt_of_ne_zero (by fun_prop) (by simp [hz₀]),
+    ← finsum_mem_support]
+  set B := closedBall (0 : ℂ) 1
+  -- separate cases depending on whether z is in the open `ball 0 1` or not
+  by_cases hzBall : z ∈ ball 0 1
+  · have hzdiv1 := divisor_sub_const_self (show z ∈ B by grind [ball_subset_closedBall])
+    suffices (fun u ↦ -(divisor (· - z) B u * log ‖u‖)).support = {z} by
+      simp [this, hzdiv1, posLog_def]
+      grind [mem_ball_zero_iff, log_nonpos, norm_nonneg]
+    rw [Function.support_eq_iff]
+    constructor
+    · simp [hzdiv1]
+      grind [mem_ball_zero_iff, norm_nonneg]
+    · intro _ hu
+      rw [Set.mem_singleton_iff] at hu
+      simp [divisor_sub_const_of_ne hu]
+  · have h1lez : 1 ≤ ‖z‖ := by grind [mem_ball_zero_iff]
+    suffices (fun u ↦ -(divisor (· - z) B u * log ‖u‖)).support = ∅ by
+      simp [this, posLog_eq_log_max_one <| norm_nonneg z, h1lez]
+    rw [Function.support_eq_empty_iff]
+    ext x
+    rcases eq_or_ne x z with rfl | hx
+    · by_cases hB : x ∈ B
+      · simp
+        grind [mem_closedBall_zero_iff]
+      · simpa using .inl <| Function.locallyFinsuppWithin.apply_eq_zero_of_notMem _ hB
+    simpa using .inl <| divisor_sub_const_of_ne hx
+
+@[simp]
+theorem logMahlerMeasure_X_add_C (z : ℂ) : (X + C z).logMahlerMeasure = log⁺ ‖z‖ := by
+  simp [← sub_neg_eq_add, ← map_neg]
+
+theorem logMahlerMeasure_C_mul_X_add_C {a : ℂ} (ha : a ≠ 0) (b : ℂ) :
+    (C a * X + C b).logMahlerMeasure = log ‖a‖ + log⁺ ‖a⁻¹ * b‖ := by
+  rw [show C a * X + C b = C a * (X + C (a⁻¹ * b)) by simp [mul_add, ← map_mul, ha],
+    logMahlerMeasure_C_mul ha (X_add_C_ne_zero (a⁻¹ * b)), logMahlerMeasure_X_add_C]
+
+theorem logMahlerMeasure_degree_eq_one {p : ℂ[X]} (h : p.degree = 1) : p.logMahlerMeasure =
+    log ‖p.coeff 1‖ + log⁺ ‖(p.coeff 1)⁻¹ * p.coeff 0‖ := by
+  rw [eq_X_add_C_of_degree_le_one (le_of_eq h)]
+  simp [logMahlerMeasure_C_mul_X_add_C (show p.coeff 1 ≠ 0 by exact coeff_ne_zero_of_eq_degree h)]
+
+/-- The Mahler measure of `X - C z` equals `max 1 ‖z‖`. -/
+@[simp]
+theorem mahlerMeasure_X_sub_C (z : ℂ) : (X - C z).mahlerMeasure = max 1 ‖z‖ := by
+  have := logMahlerMeasure_X_sub_C z
+  rw [logMahlerMeasure_eq_log_MahlerMeasure] at this
+  apply_fun exp at this
+  rwa [posLog_eq_log_max_one (norm_nonneg z),
+    exp_log (mahlerMeasure_pos_of_ne_zero <| X_sub_C_ne_zero z),
+    exp_log (lt_of_lt_of_le zero_lt_one <| le_max_left 1 ‖z‖)] at this
+
+@[simp]
+theorem mahlerMeasure_X_add_C (z : ℂ) : (X + C z).mahlerMeasure = max 1 ‖z‖ := by
+  simp [← sub_neg_eq_add, ← map_neg]
+
+theorem mahlerMeasure_C_mul_X_add_C {a : ℂ} (ha : a ≠ 0) (b : ℂ) :
+    (C a * X + C b).mahlerMeasure = ‖a‖ * max 1 ‖a⁻¹ * b‖ := by
```

Comment:
Would it make sense to have a version saying `... = max ‖a‖ ‖b‖`? This looks a bit simpler and might be useful in applications.